include/dirac_quda.h

```diff
@@ -2238,6 +2288,95 @@ namespace quda {
     }
   };
 
+  /**
+     Gloms onto a DiracMatrix and provides an operator() for its G5M method
+  */
+  class DiracG5M : public DiracMatrix
+  {
+
+  public:
+    DiracG5M(const Dirac &d) : DiracMatrix(d) { }
+    DiracG5M(const Dirac *d) : DiracMatrix(d) { }
+
+    /**
+      @brief Left-apply gamma5 as appropriate for the operator
+
+      @param vec[in,out] vector to which gamma5 is applied in place
+    */
+    void applyGamma5(ColorSpinorField &vec) const
+    {
+      // FIXME: look more carefully at if op is preconditioned or not,
+      // how it's preconditioned (symmetric vs asymmetric), etc
+      if (isWilsonType()) {
+        gamma5(vec, vec);
+      } else if (isStaggered()) {
+        // Gamma5 is (-1)^(x+y+z+t)
+        blas::ax(-1.0, vec.Odd());
+      } else if (isDwf()) {
+        // needs 5th dimension reversal, Mobius needs that inversion...
+        errorQuda("Support for Hermitian DWF operators does not exist yet");
+      } else if (isCoarse()) {
+        // more complicated, need to see if it's a repeated coarsening
+        // of the coarse op
+        errorQuda("Support for Hermitian coarse operators does not exist yet");
+      } else {
+        errorQuda("Unexpected operator type");
+      }
+    }
+
+    void operator()(ColorSpinorField &out, const ColorSpinorField &in) const
+    {
+      dirac->M(out, in);
+      if (shift != 0.0) blas::axpy(shift, const_cast<ColorSpinorField &>(in), out);
+      applyGamma5(out);
+    }
+
+    void operator()(ColorSpinorField &out, const ColorSpinorField &in, ColorSpinorField &tmp) const
+    {
+      bool reset1 = false;
+      if (!dirac->tmp1) {
+        dirac->tmp1 = &tmp;
+        reset1 = true;
+      }
+      dirac->M(out, in);
+      if (shift != 0.0) blas::axpy(shift, const_cast<ColorSpinorField &>(in), out);
+      applyGamma5(out);
+      if (reset1) {
+        dirac->tmp1 = NULL;
+        reset1 = false;
+      }
+    }
+
+    void operator()(ColorSpinorField &out, const ColorSpinorField &in, ColorSpinorField &Tmp1, ColorSpinorField &Tmp2) const
+    {
+      bool reset1 = false;
+      bool reset2 = false;
+      if (!dirac->tmp1) {
+        dirac->tmp1 = &Tmp1;
+        reset1 = true;
+      }
+      if (!dirac->tmp2) {
+        dirac->tmp2 = &Tmp2;
+        reset2 = true;
+      }
+      dirac->M(out, in);
+      if (shift != 0.0) blas::axpy(shift, const_cast<ColorSpinorField &>(in), out);
+      applyGamma5(out);
+      if (reset2) {
+        dirac->tmp2 = NULL;
+        reset2 = false;
+      }
+      if (reset1) {
+        dirac->tmp1 = NULL;
+        reset1 = false;
+      }
+    }
+
+    int getStencilSteps() const { return dirac->getStencilSteps(); }
+
+    virtual bool hermitian() const { return true; } // gamma5 op is always Hermitian
```

Comment:
FYI, removing this: it's not necessarily Hermitian, and even if so it's not strictly HPD. This also hits issues in the eigensolver class, where the Hermitian check enforces smallest real and largest real only, which means you can't extract the interior spectrum of a gamma5 operator.

Reply:
Well, correction, modifying it, because some combinations are still technically HPD.